storm-client/src/jvm/org/apache/storm/executor/Executor.java

```diff
@@ -270,6 +273,15 @@ public ExecutorShutdown execute() throws Exception {
 
     @Override
     public void accept(Object event) {
+        if (this.needToRefreshCreds) {
+            this.needToRefreshCreds = false;
```

Comment:
There seems to be a race condition here.

The executor thread consumes from receiveQ and every time it consumes one tuple, the `accept` method is invoked and `needToRefreshCreds` is checked.

This `needToRefreshCreds` variable is accessed by others threads (which invoke `credentialsChanged`) and the value will be updated to `true`.

So first of all, `needToRefreshCreds` needs to be at least `violate`. But `violate` doesn't seem enough because `needToRefreshCreds` is set to `false` in this executor thread, and set to `true` in other threads.

Between checking if `needToRefreshCreds` is true and setting it to be `false` in the executor thread, `needToRefreshCreds` could be modified to `true` in other threads but this information is lost.

Reply:
The _needToRefreshCreds_ could be updated to _true_ anytime after if condition is evaluated by the worker credentials refresh thread. If there are consecutive updates, it will either execute the _setCredentials_ twice or will execute only once with latest credentials - no harm/no deadlock.